# Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto

## Intro

### The purpose of this assignment is to explore, segment, and cluster the neighborhoods in the city of Toronto.

In order to begin the exploration of the neighborhoods, the data has to be collected, stored and cleaned.
The wikipedia page: [List of postal codes of Canada: M](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) contains a list of postal codes with the first letter M. These postal codes are located within the city of Toronto. The table provided will serve as our source.

## Import Libraries

In [1]:
import numpy as np
import pandas as pd

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

import requests

import matplotlib.cm as cm

import matplotlib.colors as colors

from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium


Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.



## Data Collection

### Method 1: Excel file

The wiki table was transferred to an excel table in order to read it easier into a dataframe

In [2]:
df = pd.read_excel(r'/Users/odettecalderon/Desktop/Neighborhoods_Toronto_Excel.xlsx')

In [3]:
print(df.shape)
df.head(7)

(180, 3)


Postal Code           Borough                                 Neighborhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Method 2: Pandas read_html function

In this method, we use the pandas read_html function to read the url where the table is found and store the tables found into a dataframe.

In [4]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
dfs = pd.read_html(url)

print(len(dfs))

3


The length indicates that there were three different tables found in the url. By indicating which table we want to see, we can store the dataframe separately. We can see that the relevant table is found under dfs[0] and we can therefore store this table into a different variable 

In [5]:
dfs[0].head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [6]:
dfs_0 = dfs[0]
dfs_0.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

## Data Cleanup

For simplicity the df generated from the excel file will be used for further analysis

Based on the assignment, the following conditions have to be met: 
- Only  the cells that have an assigned borough will be processed. Cells with a borough that is Not assigned will be ignored.
- More than one neighborhood can exist in one postal code area. The postal code will consist of a single row with the neighborhoods separated with a comma.
- If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

To have an idea of how many boroughs will be discarded, we can use value_counts to see the amount of instances that 'Not assigned' appears in the column 'Borough'

In [7]:
df['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East Toronto         5
East York            5
York                 5
Mississauga          1
Name: Borough, dtype: int64

Checking to see the condition:
- If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [8]:
df[df['Neighborhood'] == "Not assigned"]['Borough'].value_counts()

Not assigned    77
Name: Borough, dtype: int64

By filtering out the neighborhoods with 'Not assigned' and checking the unique values of the boroughs, we are able to see that all Not assigned neighborhoods have a not assigned borough and can all therefore be discarded.

We will create a new dataframe where the condition has to be met that 'not assigned' is not found in the column Borough, therefore deleting these entries

In [9]:
toronto_df = df[df['Borough'] != 'Not assigned'].reset_index(drop=True)

In [10]:
toronto_df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Dataframe shape:

In [11]:
toronto_df.shape

(103, 3)

### Obtain the latitude and longitude values for our PostalCodes

An attempt was made to obtain the lat and long for each of the postal codes in our Toronto Dataframe using geocoder

#### Defining a function that requires the postal code as an input and returns the latitude and longitude values.

In [12]:
import geocoder
geolocator = Nominatim(user_agent = 'my-application')
def get_coordinates(postal_code):
    address = postal_code + ', Toronto, Ontario'
    location = geolocator.geocode(address)
    if not location:
        return "N/A"
    else:
        latitude = location.latitude
        longitude = location.longitude
        return(latitude,longitude)

#### This function takes a list as an input and goes through every list item inputting the result into the formerly created latitude and longitude list in order.

In [13]:
latitude_list = []
longitude_list = []
def lat_lon_lists(neighborhood_list):
    for i in neighborhood_list:
        coordinate = get_coordinates(i)
        if coordinate == "N/A":
            latitude_list.append("N/A")
            longitude_list.append("N/A")
        else:
            latitude_list.append(coordinate[0])
            longitude_list.append(coordinate[1])

Converting the Postal Code column into a list to input into the function

In [14]:
postalcode_list = toronto_df['Postal Code'].tolist()

#### Calling the function on the postal code list

In [15]:
lat_lon_lists(postalcode_list)

Seeing the results: As observed the list contains a lot of not available information. By checking the size of the list we can observe that the functions succesfully went through all values in our Postal Code column which gives us a total of 103 list entries

In [16]:
print(longitude_list[0:10])

[-79.3839347, 'N/A', 'N/A', 'N/A', -79.3839347, 'N/A', -79.3839347, 'N/A', 'N/A', 'N/A']


In [17]:
print(len(longitude_list), len(latitude_list))

103 103


## Lat and longitude values using google geocoder

testing google geocoder on a single Postal Code (M4A) whose coordinates we were not able to obtain earlier

In [19]:
import geocoder
tor_address = 'M4A, Toronto, Ontario'
lat_long_coords = None
while lat_long_coords == None:
    g = geocoder.google(tor_address)
    lat_long_coords = g.latlng

KeyboardInterrupt: 

Took too long and was manually interruped

In [21]:
print(lat_long_coords)

None


## Lat and longitude values using csv file

Instead of using geopy, we will now use the provided csv file to obtain the latitude and longitude coordinates of our Postal Codes. We begin by using read_csv to store the file into a dataframe. The shape and preview using .head will give us an idea of what we are dealing with

In [22]:
lat_long_df = pd.read_csv('/Users/odettecalderon/Downloads/Geospatial_Coordinates.csv')
print(lat_long_df.shape)
lat_long_df.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

As we can see, the provided dataframe is sorted by PostalCode. Currently, our Toronto dataframe is not sorted.

In [23]:
toronto_df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In order to facilitate the union of the two Dataframes, we will sort the toronto_df. We will have to reset the index again as well.

In [24]:
toronto_df.sort_values(by=['Postal Code'],inplace= True)
toronto_df.reset_index(drop=True,inplace= True)
toronto_df.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

The pandas function for performing joins is called merge and an Inner join is the default option:
Inner joins yield a DataFrame that contains only rows where the value being joined exists in BOTH tables

In [25]:
toronto_df = pd.merge(toronto_df,lat_long_df)
print(toronto_df.shape)
toronto_df.head()

(103, 5)


Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Creating a Map of Toronto with Neighborhoods Superimposed

Getting geo coordinates of Toronto 

In [26]:
location_tor = geolocator.geocode('Toronto, Ontario, Canada')
latitude_tor = location_tor.latitude
longitude_tor = location_tor.longitude
print('The coordinates of Toronto are {}, {}'.format(latitude_tor,longitude_tor))

The coordinates of Toronto are 43.6534817, -79.3839347


### Visualization of all Neighborhoods using Folium

In [27]:
map_toronto = folium.Map(location = [latitude_tor, longitude_tor],zoom_start = 10)
for lat,lng, borough,neighborhood in zip(toronto_df['Latitude'],
                                         toronto_df['Longitude'],
                                         toronto_df['Borough'],
                                         toronto_df['Neighborhood']):
    label = '{}.{}'.format(neighborhood,borough)
    label = folium.Popup(label,parse_html = True)
    folium.CircleMarker([lat,lng],radius = 4,
                        popup = label,
                        fill = True,
                        color = 'Blue',
                        fill_color = '#3186cc',
                        fill_opacity = 0.9,
                      parse_html = False).add_to(map_toronto)

In [28]:
map_toronto

# Exploration of Boroughs with *Toronto* in their name

In [29]:
toronto_df['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
East Toronto         5
York                 5
Mississauga          1
Name: Borough, dtype: int64

### Obtain all rows in dataframe that contain a given substring

In [30]:
toronto_data = toronto_df[toronto_df['Borough'].str.contains('Toronto')].reset_index(drop=True)

In [31]:
print(toronto_data.shape,'is equal to amount seen in value_counts')
toronto_data.head()

(39, 5) is equal to amount seen in value_counts


Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

### Foursquare Credentials

In [32]:
from IPython.display import HTML
from IPython.display import display

# Taken from https://stackoverflow.com/questions/31517194/how-to-hide-one-specific-cell-input-or-output-in-ipython-notebook
tag = HTML('''<script>
code_show=true; 
function code_toggle() {
    if (code_show){
        $('div.cell.code_cell.rendered.selected div.input').hide();
    } else {
        $('div.cell.code_cell.rendered.selected div.input').show();
    }
    code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
Raw code input hidden, click <a href="javascript:code_toggle()">here</a> to see.''')
display(tag)
CLIENT_ID = 'LE2M35FUCSB4CCTN3AOU5L3HYJ5LVPNR2WQUO1RXBXETFMT2'
CLIENT_SECRET = 'KEPWPFIGYYQZ5LRVNUATDT4HM10B04HBABHKA0EKPOCA30BU'
VERSION = '20180605'

Define Function *getNearbyvenues* that obtains the latitude and longitude values from dataframe, sends the get request, extracts the top 100 venues (within 500 meters) and their respective categories

In [33]:
def getNearbyVenues(names, latitudes, longitudes, radius=500,limit=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Code to run the function on each neighborhood and create new dataframe called toronto_venues

In [34]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                 longitudes=toronto_data['Longitude'],
                                 latitudes=toronto_data['Latitude'])
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                 Seaspray Restaurant       43.678888       -79.298167   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Asian Restaurant

In [35]:
print('Shape:',toronto_venues.shape)
toronto_venues.tail()

Shape: (1620, 7)


Neighborhood  \
1615  Business reply mail Processing Centre, South C...   
1616  Business reply mail Processing Centre, South C...   
1617  Business reply mail Processing Centre, South C...   
1618  Business reply mail Processing Centre, South C...   
1619  Business reply mail Processing Centre, South C...   

      Neighborhood Latitude  Neighborhood Longitude                    Venue  \
1615              43.662744              -79.321558     TTC Russell Division   
1616              43.662744              -79.321558  Jonathan Ashbridge Park   
1617              43.662744              -79.321558             The Ten Spot   
1618              43.662744              -79.321558         Olliffe On Queen   
1619              43.662744              -79.321558              ONE Academy   

      Venue Latitude  Venue Longitude        Venue Category  
1615       43.664908       -79.322560    Light Rail Station  
1616       43.664702       -79.319898                  Park  
1617       43.664815       -79.324213                   Spa  
1618       43.664503       -79.324768               Butcher  
1619       43.662253       -79.326911  Gym / Fitness Center

Venues grouped by Neighborhood

In [36]:
toronto_venues.groupby('Neighborhood').count().head()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           24   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     64   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            24   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      64   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            56              56   
Brockton, Parkdale Village, Exhibition Place           24              24   
Business reply mail Processing Centre, South Ce...     16              16   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     64              64   

                                                    Venue Longitude  \
Neighborhood                                                          
Berczy Park                                                      56   
Brockton, Parkdale Village, Exhibition Place                     24   
Business reply mail Processing Centre, South Ce...               16   
CN Tower, King and Spadina, Railway Lands, Harb...               16   
Central Bay Street                                               64   

                                                    Venue Category  
Neighborhood                                                        
Berczy Park                                                     56  
Brockton, Parkdale Village, Exhibition Place                    24  
Business reply mail Processing Centre, South Ce...              16  
CN Tower, King and Spadina, Railway Lands, Harb...              16  
Central Bay Street                                              64

Amount of unique venue categories:

In [37]:
print('There are {} unique categories of venues'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 unique categories of venues


# Analyzing each neighborhood

In [38]:
#one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
#drop neighbordhood and add column back to dataframe
toronto_onehot.drop('Neighborhood',axis=1,inplace=True)
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']
#move neighborhood column to first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0  The Beaches                  0        0                   0             0   
1  The Beaches                  0        0                   0             0   
2  The Beaches                  0        0                   0             0   
3  The Beaches                  0        0                   0             0   
4  The Beaches                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Toy / Game Store  Trail  Train Station  \
0             0  ...                 0      1              0   
1             0  ...                 0      0              0   
2             0  ...                 0      0              0   
3             0  ...                 0      0              0   
4             0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0         0          0              0            0  
1         0          0              0            0  
2         0          0              0            0  
3         0          0              0            0  
4         0          0              0            0  

[5 rows x 237 columns]

In [39]:
toronto_onehot.shape

(1620, 237)

Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [40]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0   0.0000              0.0000        0.0000           0.000           0.0000   
1   0.0000              0.0000        0.0000           0.000           0.0000   
2   0.0000              0.0000        0.0000           0.000           0.0000   
3   0.0625              0.0625        0.0625           0.125           0.1875   
4   0.0000              0.0000        0.0000           0.000           0.0000   

   Airport Terminal  American Restaurant  Antique Shop  ...  Toy / Game Store  \
0             0.000                  0.0           0.0  ...               0.0   
1             0.000                  0.0           0.0  ...               0.0   
2             0.000                  0.0           0.0  ...               0.0   
3             0.125                  0.0           0.0  ...               0.0   
4             0.000                  0.0           0.0  ...               0.0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0    0.0            0.0                       0.017857               0.0   
1    0.0            0.0                       0.000000               0.0   
2    0.0            0.0                       0.000000               0.0   
3    0.0            0.0                       0.000000               0.0   
4    0.0            0.0                       0.015625               0.0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0                    0.0  0.000000        0.0            0.0     0.000000  
1                    0.0  0.000000        0.0            0.0     0.041667  
2                    0.0  0.000000        0.0            0.0     0.000000  
3                    0.0  0.000000        0.0            0.0     0.000000  
4                    0.0  0.015625        0.0            0.0     0.015625  

[5 rows x 237 columns]

In [41]:
toronto_grouped.shape

(39, 237)

### Print of each neighborhood along with the top 5 most common venues

In [42]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2                Café  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1     Coffee Shop  0.08
2  Breakfast Spot  0.08
3          Bakery  0.08
4     Yoga Studio  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
           venue  freq
0    Pizza Place  0.06
1  Auto Workshop  0.06
2     Restaurant  0.06
3        Butcher  0.06
4  Burrito Place  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0      Airport Service  0.19
1       Airport Lounge  0.12
2     Airport Terminal  0.12
3  Rental Car Location  0.06
4              Airport  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1 

### Function to sort the venues in descending order.

In [43]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [44]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :],
                                                                          num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar                Bakery   
1                Bakery           Coffee Shop        Breakfast Spot   
2           Pizza Place            Skate Park               Brewery   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0    Seafood Restaurant           Cheese Shop             Restaurant   
1           Yoga Studio     Convenience Store  Performing Arts Venue   
2                   Spa         Burrito Place         Farmers Market   
3               Airport    Airport Food Court           Airport Gate   
4   Japanese Restaurant       Thai Restaurant            Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café          Gourmet Shop                    Pub  
1             Pet Store          Climbing Gym             Restaurant  
2  Fast Food Restaurant            Restaurant                Butcher  
3              Boutique         Boat or Ferry    Rental Car Location  
4       Bubble Tea Shop          Burger Joint       Department Store

## Cluster Neighborhoods

In [45]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 1], dtype=int32)

#### New dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [46]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [47]:
toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0     Health Food Store      Asian Restaurant   
1 -79.352188               2      Greek Restaurant           Coffee Shop   
2 -79.315572               2                  Park  Fast Food Restaurant   
3 -79.340923               2                  Café           Coffee Shop   
4 -79.388790               1                  Park              Bus Line   

  3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0                   Pub                   Trail                 Diner   
1    Italian Restaurant  Furniture / Home Store             Bookstore   
2           Pizza Place              Board Shop     Fish & Chips Shop   
3             Gastropub                  Bakery               Brewery   
4           Swim School                  Lawyer    College Rec Center   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Deli / Bodega      Department Store          Dessert Shop   
1        Ice Cream Shop            Restaurant          Liquor Store   
2               Brewery            Restaurant    Italian Restaurant   
3   American Restaurant             Bookstore        Sandwich Place   
4    College Auditorium   Dumpling Restaurant            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
0    Dim Sum Restaurant            Yoga Studio  
1     Indian Restaurant              Juice Bar  
2         Burrito Place           Intersection  
3           Cheese Shop              Pet Store  
4      Doner Restaurant                Dog Run

### Visualization of Clusters

In [48]:
map_clusters = folium.Map(location=[latitude_tor, longitude_tor], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], 
                                  toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine Clusters

Examining each cluster and determining the discriminating venue categories that distinguish each cluster.

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + 
                                                                list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  East Toronto               0     Health Food Store      Asian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub                 Trail                 Diner   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Deli / Bodega      Department Store          Dessert Shop   

  9th Most Common Venue 10th Most Common Venue  
0    Dim Sum Restaurant            Yoga Studio

In [50]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + 
                                                                list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4   Central Toronto               1                  Park   
5   Central Toronto               1                  Park   
23  Central Toronto               1                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4               Bus Line           Swim School                Lawyer   
5      Convenience Store        Breakfast Spot        Sandwich Place   
23                  Park              Bus Line         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4     College Rec Center    College Auditorium   Dumpling Restaurant   
5       Department Store     Food & Drink Shop                 Hotel   
23      Sushi Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4             Donut Shop      Doner Restaurant                Dog Run  
5                    Gym           Pizza Place          Deli / Bodega  
23      Doner Restaurant               Dog Run    Distribution Center

It is interesting to note that the most common venues within Cluster 1 is a Park which gives the impression that these areas are greener than other parts of the city

In [51]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + 
                                                                list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       East Toronto               2      Greek Restaurant   
2       East Toronto               2                  Park   
3       East Toronto               2                  Café   
6    Central Toronto               2        Clothing Store   
7    Central Toronto               2        Sandwich Place   
9    Central Toronto               2           Coffee Shop   
11  Downtown Toronto               2                  Café   
12  Downtown Toronto               2           Coffee Shop   
13  Downtown Toronto               2           Coffee Shop   
14  Downtown Toronto               2        Clothing Store   
15  Downtown Toronto               2                  Café   
16  Downtown Toronto               2           Coffee Shop   
17  Downtown Toronto               2           Coffee Shop   
18  Downtown Toronto               2           Coffee Shop   
19  Downtown Toronto               2           Coffee Shop   
20  Downtown Toronto               2           Coffee Shop   
21  Downtown Toronto               2           Coffee Shop   
24   Central Toronto               2                  Café   
25  Downtown Toronto               2                  Café   
26  Downtown Toronto               2                  Café   
27  Downtown Toronto               2       Airport Service   
28  Downtown Toronto               2           Coffee Shop   
29  Downtown Toronto               2           Coffee Shop   
30  Downtown Toronto               2         Grocery Store   
31      West Toronto               2                Bakery   
32      West Toronto               2                   Bar   
33      West Toronto               2                  Café   
34      West Toronto               2    Mexican Restaurant   
35      West Toronto               2        Breakfast Spot   
36      West Toronto               2                  Café   
37  Downtown Toronto               2           Coffee Shop   
38      East Toronto               2                  Park   

            2nd Most Common Venue          3rd Most Common Venue  \
1                     Coffee Shop             Italian Restaurant   
2            Fast Food Restaurant                    Pizza Place   
3                     Coffee Shop                      Gastropub   
6                     Coffee Shop                    Yoga Studio   
7                    Dessert Shop                    Pizza Place   
9                             Pub            Fried Chicken Joint   
11                    Coffee Shop                            Pub   
12               Sushi Restaurant            Japanese Restaurant   
13                           Park                            Pub   
14                    Coffee Shop             Italian Restaurant   
15                    Coffee Shop                     Restaurant   
16                   Cocktail Bar                       Beer Bar   
17                 Sandwich Place             Italian Restaurant   
18                           Café                     Restaurant   
19                       Aquarium                           Café   
20                          Hotel                           Café   
21                     Restaurant                           Café   
24                 Sandwich Place                    Coffee Shop   
25             Italian Restaurant                            Bar   
26  Vegetarian / Vegan Restaurant                         Bakery   
27                 Airport Lounge               Airport Terminal   
28                           Café                     Restaurant   
29                           Café                     Restaurant   
30                           Café                           Park   
31                       Pharmacy                           Bank   
32                     Restaurant  Vegetarian / Vegan Restaurant   
33                         Bakery                    Coffee Shop   
34                Thai Restaurant                           Ca

In [52]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + 
                                                                list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
22           Yoga Studio          Dance Studio  Eastern European Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22   Dumpling Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22               Dog Run   Distribution Center         Discount Store

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + 
                                                                list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8    Central Toronto               4                  Park   
10  Downtown Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8             Restaurant            Playground      Doner Restaurant   
10            Playground                 Trail      Cuban Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8                Dog Run   Distribution Center        Discount Store   
10            Donut Shop      Doner Restaurant               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8                  Diner    Dim Sum Restaurant           Dessert Shop  
10   Distribution Center        Discount Store                  Diner